src/librustc_lint/unused.rs

```diff
@@ -103,11 +103,33 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
                         op_warned = true;
                     },
-                    _ => {},
+                    hir::BiAdd | hir::BiSub | hir::BiDiv | hir::BiMul | hir::BiRem => {
+                        let msg = "unused arithmetic operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
+                    },
+                    hir::BiAnd | hir::BiOr => {
+                        let msg = "unused logical operation which must be used";
+                        cx.span_lint(UNUSED_MUST_USE, expr.span, msg);
+                        op_warned = true;
```

Comment:
For bonus points you could emit a help message here saying to use `if (first_op) { second_op }` instead, but that's a bit involved. I might tack that on myself if you don't want to get into it.